### PR TITLE
[mlir][spirv] Update math.powf lowering

### PR DESCRIPTION
**Pull request by d-smirnov**
The PR updates math.powf lowering to produce NaN result for a negative base with a fractional exponent which matches the actual behaviour of the C/C++ implementation.

### CODE REVIEW

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp [hunk 1]
```diff
@@ -389,7 +389,31 @@ struct PowFOpPattern final : public OpConversionPattern<math::PowFOp> {
...
+    // Calculae calc reminder from exponent and check whether it is zero
+    Value floatOne =
+        spirv::ConstantOp::getOne(adaptor.getRhs().getType(), loc, rewriter);
+    Value expRem =
+        rewriter.create<spirv::FRemOp>(loc, adaptor.getRhs(), floatOne);
+    Value expRemNonZero =
+        rewriter.create<spirv::FOrdNotEqualOp>(loc, expRem, zero);
+    Value cmpNegativeWithFractionalExp =
+        rewriter.create<spirv::LogicalAndOp>(loc, expRemNonZero, lessThan);
+    // Create NaN result and replace base value if conditions meet
```
Comment (kuhar) [inline mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp:405]:
```suggestion
    // Create NaN result and replace base value if condition are met.
```

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp [hunk 2]
```diff
@@ -389,7 +389,31 @@ struct PowFOpPattern final : public OpConversionPattern<math::PowFOp> {
...
     Value lessThan =
         rewriter.create<spirv::FOrdLessThanOp>(loc, adaptor.getLhs(), zero);
-    Value abs = rewriter.create<spirv::GLFAbsOp>(loc, adaptor.getLhs());
+
+    // Per C/CPP spec:
+    // "pow(base, exponent) returns NaN (and raises FE_INVALID) if base is "
+    // " finite and negative and exponent is finite and non-integer. "
+    // Calculae calc reminder from exponent and check whether it is zero
+    Value floatOne =
+        spirv::ConstantOp::getOne(adaptor.getRhs().getType(), loc, rewriter);
```
Comment (kuhar) [inline mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp:398]: Can you use a local variable for operand type?

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp [hunk 3]
```diff
@@ -389,7 +389,31 @@ struct PowFOpPattern final : public OpConversionPattern<math::PowFOp> {
...
+    Value expRem =
+        rewriter.create<spirv::FRemOp>(loc, adaptor.getRhs(), floatOne);
+    Value expRemNonZero =
+        rewriter.create<spirv::FOrdNotEqualOp>(loc, expRem, zero);
+    Value cmpNegativeWithFractionalExp =
+        rewriter.create<spirv::LogicalAndOp>(loc, expRemNonZero, lessThan);
+    // Create NaN result and replace base value if conditions meet
+    const auto &floatSemantics = scalarFloatType.getFloatSemantics();
+    const auto nan = APFloat::getNaN(floatSemantics);
```
Comment (kuhar) [inline mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp:407]: Is this an issue for webgpu? @ScottTodd

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp [hunk 4]
```diff
@@ -377,40 +377,40 @@ struct PowFOpPattern final : public OpConversionPattern<math::PowFOp> {
...
       auto shape = vectorType.getShape();
       intType = VectorType::get(shape, scalarIntType);
     }
 
     // Per GL Pow extended instruction spec:
     // "Result is undefined if x < 0. Result is undefined if x = 0 and y <= 0."
     Location loc = powfOp.getLoc();
-    Value zero =
-        spirv::ConstantOp::getZero(adaptor.getLhs().getType(), loc, rewriter);
+    auto baseType = adaptor.getLhs().getType();
```
Comment (kuhar) [inline mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp:389]: Isn't LHS required to be the same as the RHS? https://mlir.llvm.org/docs/Dialects/MathOps/#mathpowf-mathpowfop

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp [hunk 5]
```diff
@@ -377,40 +377,40 @@ struct PowFOpPattern final : public OpConversionPattern<math::PowFOp> {
...
+    // Calculate the reminder from the exponent and check whether it is zero.
+    Value floatOne = spirv::ConstantOp::getOne(exponentType, loc, rewriter);
     Value expRem =
         rewriter.create<spirv::FRemOp>(loc, adaptor.getRhs(), floatOne);
     Value expRemNonZero =
         rewriter.create<spirv::FOrdNotEqualOp>(loc, expRem, zero);
     Value cmpNegativeWithFractionalExp =
         rewriter.create<spirv::LogicalAndOp>(loc, expRemNonZero, lessThan);
-    // Create NaN result and replace base value if conditions meet
+    // Create NaN result and replace base value if conditions are met
```
Comment (kuhar) [inline mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp:405]:
```suggestion
    // Create NaN result and replace base value if conditions are met.
```

**Review by kuhar — APPROVED** [review]

**Review by GeorgeARM — APPROVED** [review]

**Pull request merged**